### DISCUSSION TITLE
App choice at compile time

### DISCUSSION THREAD

**Discussion by bbadrignans** [InfiniTimeOrg/InfiniTime#2062] [Q&A] [answered]
The main branch seems to have support for choosing apps at compile time. But I can't figure out how. 
Is it yet support ? If yes how ?

**Comment by FintasticMan** [comment, marked as answer]
It is supported! You can add a flag to CMake to choose which apps get compiled into the firmware: `-DENABLE_USERAPPS="<apps you want here>"`. See the end of [this document](https://github.com/InfiniTimeOrg/InfiniTime/blob/main/doc/code/Apps.md).

**Reply by bbadrignans** [reply]
Thanks and sorry for the RTFM question

**Reply by FintasticMan** [reply]
No problem

**Reply by bbadrignans** [reply]
I'm trying to add an app (see here if you want : https://github.com/bbadrignans/InfiniTime/tree/feat/home).
I cannot compile due to : 
```
$ cmake -DARM_NONE_EABI_TOOLCHAIN_PATH=[...]/gcc-arm-none-eabi-10.3-2021.10/ -DNRF5_SDK_PATH=[...]/nRF5_SDK_15.3.0_59ac345/ -DCMAKE_BUILD_TYPE=Release -DBUILD_DFU=0 -DBUILD_RESOURCES=1 -DTARGET_DEVICE=PINETIME
$ make -j4 pinetime-mcuboot-app
[...]
gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld: region RAM overflowed with stack
Memory region         Used Size  Region Size  %age Used
           FLASH:      314344 B     474632 B     66.23%
     SPARE_SPACE:          0 GB        12 KB      0.00%
             RAM:       64520 B        64 KB     98.45%
collect2: error: ld returned 1 exit status
make[3]: *** [src/CMakeFiles/pinetime-mcuboot-app.dir/build.make:1990 : src/pinetime-mcuboot-app-1.14.0.out] Erreur 1
make[2]: *** [CMakeFiles/Makefile2:316 : src/CMakeFiles/pinetime-mcuboot-app.dir/all] Erreur 2
make[1]: *** [CMakeFiles/Makefile2:323 : src/CMakeFiles/pinetime-mcuboot-app.dir/rule] Erreur 2
make: *** [Makefile:189 : pinetime-mcuboot-app] Erreur 2
```
I can compile last year with the same app.
I tried to remove other apps (Music, Navigation ...) and watchfaces, but it does not change RAM usage a lot (few hundred of bytes).
Am I doing things wrong ?

**Reply by JF002** [reply]
Since last year, InfiniTime statically allocates most of the available RAM memory to make the [FreeRTOS heap](https://github.com/InfiniTimeOrg/InfiniTime/blob/main/src/FreeRTOSConfig.h#L65) as big as possible. This heap is used for all dynamic allocation at run-time.

If your apps needs to allocate static memory, you'll probably need to reduce the size of the heap accordingly.

**Reply by bbadrignans** [reply]
@JF002 Thanks a lot for your help.
My app doesn't need so much memory but it was necessary.

One other question, does something change since last year regarding Bluetooth ?
My app can't send Bluetooth packets when a LVGL button is pressed. Worked fine last year.
I checked with Wireshark : I can see packets from Music app buttons, but not from mine.
My app is mostly based on Music app, except it uses also MotorController.
I triple check that my last merge in Infinitime main branch is ok [(reduce to just 2 commits)](https://github.com/bbadrignans/InfiniTime/compare/InfiniTimeOrg%3AInfiniTime%3Amain...feat/home)

Any idea ? I'm stuck with my sealed watch and without a debugger.

**Reply by JF002** [reply]
InfiniTime allocates nearly 100% of the RAM memory, so it's normal you have to free a bit of memory if you app needs a few 10s or 100s of bytes :-) 

I don't think we made any significant changes in the BLE implementation recently. Did you check that your HomeService is discovered by your applicatioin on the host side? And that it subscribed  to the notification that is sent in `Pinetime::Controllers::HomeService::event()` ?

**Reply by bbadrignans** [reply]
Thank you for your answer. 
I think I was also confused because, after configuring apps using cmake, one needs to "make clean" and/or remove CMakeCache.txt. 
So memory usage was not changing adding/removing apps.
Now I know for the "clean" required :-)

Regarding my bluetooth issue, sorry but it was a mistake on my side during merge.
I open then close this discussion https://github.com/InfiniTimeOrg/InfiniTime/discussions/2071 one hour before your answer :-)
Just forgot to register my app : 
`displayApp.Register(&nimbleController.home());`

I will try to make this app more "user friendly" and propose a PR.

Regards

**Reply by JF002** [reply]
the ~~lord~~ CMake cache works in mysterious ways... I probably clear my CMake cache more than necessary, but it saves  me from debugging strange issues :D

Happy that you found a solution to your bluetooth issue :+1
